### PR TITLE
Improve iso

### PR DESCRIPTION
**Pull request by james-d-mitchell**
This PR uses the Digraphs package isomorphism finding capabilities to find isomorphisms between semigroups (if they exist). This change improves the performance of some examples by up to 50 times. Just to mention that graph isomorphism and semigroup isomorphism are polynomially equivalent, and so it seems unlikely that we can do better than this (although there is certainly some room for improving the encoding of a semigroup in a digraph). 

This is WIP since I still have some changes to make, but want to make sure that the CI passes.

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
This is ready for review @flsmith !

File: doc/isomorph.xml [hunk 1]
```diff
@@ -96,20 +199,19 @@ false]]></Example>
...
+        between vertex coloured digraphs in &bliss; via 
+      <Ref Oper="IsomorphismDigraphs" 
+           Label="for digraphs and homogeneous lists"
+           BookName="digraphs"/> using digraphs constructed from the
+        multiplication tables of <A>S</A> and <A>T</A>.
+      <P/>
+
+      Note that finding an isomorphism between two semigroups is difficult, and
+      may not be possible for semigroups whose size exceeds approximately
+      2048 elements.  On the other hand, <C>IsomorphismSemigroups</C> may be
```
Comment (flsmith) [inline doc/isomorph.xml:212]: 2048 is rather optimistic...

File: gap/attributes/isomorph.gi [hunk 2]
```diff
@@ -171,50 +190,150 @@ end);
...
+  s := Representative(MaximalDClasses(S)[1]);
+  t := Representative(MaximalDClasses(T)[1]);
+  SS := Semigroup(s);
+  TT := Semigroup(t);
+  map := x -> t ^ Length(Factorization(SS, x));
+  inv := x -> s ^ Length(Factorization(TT, x));
+  return MagmaIsomorphismByFunctionsNC(S, T, map, inv);
 end);
 
+SEMIGROUPS.DigraphFromMultTable := function(S)
```
Comment (flsmith) [inline gap/attributes/isomorph.gi:216]: Function name isn't fool-proofed, as proved by a fool.

File: gap/attributes/isomorph.gi [hunk 3]
```diff
@@ -171,50 +190,150 @@ end);
...
+  if p = fail then
     return fail;
   fi;
+  p := RestrictedPerm(p, [1 .. Size(S)]);
+  map := x -> AsSSortedList(T)[PositionSorted(S, x) ^ p];
+  inv := x -> AsSSortedList(S)[PositionSorted(T, x) ^ (p ^ -1)];
+  return MagmaIsomorphismByFunctionsNC(S, T, map, inv);
+end);
+
+InstallMethod(AutomorphismGroup, "for a semigroup",
```
Comment (flsmith) [inline gap/attributes/isomorph.gi:310]: I wonder if it's possible to just compute the automorphism group of the cayley digraph without edge colours, then wave some group theory magic wand to restrict the automorphisms found. That way you don't have to produce gigantic digraphs (though you do need a magic wand).

File: doc/isomorph.xml (hunk 1 above)
Comment (flsmith) [reply doc/isomorph.xml:212]: I've succeeded on semigroups of size 500 with 3 generators, which used about 10GB of RAM.

File: gap/attributes/isomorph.gi [hunk 4]
```diff
@@ -171,50 +190,150 @@ end);
...
+      return IsomorphismSemigroups(S, T);
+    else
+      return fail;
+    fi;
+  elif Size(S) <> Size(T) or NrDClasses(S) <> NrDClasses(T) then
+    return fail;
+  fi;
 
   # Monogenic semigroups are of the same size, are not groups, and have the
   # same number of D-classes by this point, and so they are isomorphism
```
Comment (flsmith) [inline gap/attributes/isomorph.gi:206]: isomorphism -> isomorphic

File: scripts/travis-build-dependencies.sh [hunk 5]
```diff
@@ -58,7 +58,7 @@ CURL="curl --connect-timeout 5 --max-time 10 --retry 5 --retry-delay 0 --retry-m
 
 ################################################################################
 # Install digraphs, genss, io, orb, and profiling
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply scripts/travis-build-dependencies.sh:60]: Done

File: tst/standard/isomorph.tst [hunk 6]
```diff
@@ -222,6 +233,153 @@ gap> T := MonogenicSemigroup(3, 1);
...
+ , function( x ) ... end, function( x ) ... end )
+
+# AutomorphismGroup
+gap> S := JonesMonoid(5);
+<regular bipartition *-monoid of degree 5 with 4 generators>
+gap> G := AutomorphismGroup(S);
+<group with 1 generators>
+gap> StructureDescription(G):
+Syntax error: ; expected in stream:1
+StructureDescription(G):
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply tst/standard/isomorph.tst:374]: Fixed

File: doc/isomorph.xml (hunk 1 above)
Comment (james-d-mitchell) [reply doc/isomorph.xml:212]: Revised downwards to "a few hundred elements"

File: gap/attributes/isomorph.gi (hunk 2 above)
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:216]: Two of us at least, renamed to "IsomorphismInvariantDigraph"

File: gap/attributes/isomorph.gi (hunk 3 above)
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:310]: I'm not sure, I tried something like this, but maybe didn't have a big enough wand. For example, for isomorphism testing, the semigroups with multiplication tables:
~~~~
1111
1111
1111
1112
~~~~
and 
~~~~
1111
1111
1111
1121
~~~~
are not isomorphic (or anti-isomorphic) and if you do `Digraph(MultiplicationTable())`, then the resulting digraphs are **equal** (according to the operator = for digraphs in GAP) not only isomorphic. I have no doubt that the method I've given here is a fallback and that there are better methods for some types of semigroups (see the comment about RZMS for example). Anyway, it's better than nothing, and I don't know any immediate ways to improve it.

File: gap/attributes/isomorph.gi (hunk 4 above)
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:206]: Fixed

File: gap/attributes/isomorph.gi [hunk 7]
```diff
@@ -171,50 +190,150 @@ end);
...
+  end;
+
+  # i = 1 .. n, j = 1 .. n
+  Color2Node := function(i, j)
+    Assert(1, 1 <= i and i <= n);
+    Assert(1, 1 <= j and j <= n);
+    return i * n + j;
+  end;
+
+  Color3Node := function(i, j)
```
Comment (wilfwilson) [inline gap/attributes/isomorph.gi:234]: I haven't worked out why these nodes need to exist yet.

File: tst/standard/isomorph.tst [hunk 8]
```diff
@@ -222,6 +233,151 @@ gap> T := MonogenicSemigroup(3, 1);
...
+      40, 40, 2, 1, 4, 3, 40, 40, 39, 39, 40, 40, 40, 39, 39, 39, 5, 6 ], 
+  [ 4, 3, 2, 1, 40, 39, 40, 39, 3, 4, 1, 2, 4, 3, 1, 2, 5, 5, 6, 6, 40, 40, 
+      39, 39, 4, 3, 2, 1, 39, 39, 40, 40, 39, 39, 39, 40, 40, 40, 6, 5 ] ]
+
+# IsomorphismSemigroups
+gap> S := FullTransformationMonoid(3);
+<full transformation monoid of degree 3>
+gap> T := AsMonoid(IsPBRMonoid, S);
+<pbr monoid of size 27, degree 3 with 3 generators>
+gap> IsomorphismSemigroups(S, T);
```
Comment (wilfwilson) [inline tst/standard/isomorph.tst:362]: A 'brute force' iso/inv test would be nice here, and with some of the other non-fail test examples, too.

File: doc/isomorph.xml [hunk 9]
```diff
@@ -1,13 +1,117 @@
...
+<#GAPDoc Label="CanonicalMultiplicationTable">
+  <ManSection>
+    <Attr Name = "CanonicalMultiplicationTable" Arg = "S"/>
+    <Returns>
+      An isomorphism invariant multiplication table of a semigroup.
+    </Returns>
+    <Description>
+      This function returns a multiplication table of a semigroup
+      isomorphic to the semigroup <A>S</A>. <C>CanonicalMultiplicationTable</C>
+      is an isomorphism invariant of semigroups, and so it can, for example, be
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply doc/isomorph.xml:80]: Aha excellent comment, you’re right I am misusing isomorphism invariant all over the place here, I’ll revise! Thanks for the review

File: gap/attributes/isomorph.gi (hunk 7 above)
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:234]: These ones actually don’t need to exist but in some experiments their presence improved the performance so I left them in.

Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:234]: These ones actually don’t need to exist but in some experiments their presence improved the performance so I left them in.

Comment (wilfwilson) [reply gap/attributes/isomorph.gi:234]: Fair enough!

File: gap/attributes/isomorph.gi [hunk 10]
```diff
@@ -171,50 +190,150 @@ end);
...
+  map := x -> AsSSortedList(T)[PositionSorted(S, x) ^ p];
+  inv := x -> AsSSortedList(S)[PositionSorted(T, x) ^ (p ^ -1)];
+  return MagmaIsomorphismByFunctionsNC(S, T, map, inv);
+end);
+
+InstallMethod(AutomorphismGroup, "for a semigroup",
+[IsSemigroup],
+function(S)
+  local D, G, X, map, Y, H;
+  # JDM: I'm not sure I trust the AutomorphismGroup methods for Rees 0-matrix
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:314]: The implementation, it needs to be overhauled and documented. I think Chris’ PR is a better way to go!

File: gap/attributes/isomorph.gi (hunk 7 above)
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi:234]: Probably this falls under the heading “the graph encoding can probably be improved”

Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi]: I actually removed the Color3Nodes now, same performance, or slightly better

File: doc/isomorph.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply doc/isomorph.xml]: Done

File: doc/isomorph.xml [hunk 11]
```diff
@@ -1,13 +1,117 @@
...
+(1,5,8,3,6,7,2,4)]]></Example>
+  </Description>
+</ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="CanonicalMultiplicationTable">
+  <ManSection>
+    <Attr Name = "CanonicalMultiplicationTable" Arg = "S"/>
+    <Returns>
+      An isomorphism invariant multiplication table of a semigroup.
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply doc/isomorph.xml]: Done

File: gap/attributes/isomorph.gi [hunk 12]
```diff
@@ -171,50 +190,150 @@ end);
...
+  s := Representative(MaximalDClasses(S)[1]);
+  t := Representative(MaximalDClasses(T)[1]);
+  SS := Semigroup(s);
+  TT := Semigroup(t);
+  map := x -> t ^ Length(Factorization(SS, x));
+  inv := x -> s ^ Length(Factorization(TT, x));
+  return MagmaIsomorphismByFunctionsNC(S, T, map, inv);
 end);
 
+SEMIGROUPS.IsomorphismInvariantDigraph := function(S)
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi]: Renamed `CanonicalDigraph`

File: gap/attributes/isomorph.gi [hunk 13]
```diff
@@ -171,50 +190,150 @@ end);
...
+  local M, n, Color1Node, Color2Node, Color3Node, Widget, out, colors, x, y, z;
+
+  M := MultiplicationTable(S);
+  n := Size(S);
+
+  # The original nodes
+  Color1Node := function(i)
+    return i;
+  end;
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi]: Done

File: gap/attributes/isomorph.gi [hunk 14]
```diff
@@ -171,50 +190,150 @@ end);
...
+  Append(colors, ListWithIdenticalEntries(n ^ 2, 2));
+  Append(colors, ListWithIdenticalEntries(n ^ 2, 3));
+  Append(colors, ListWithIdenticalEntries(n, 2));
+
+  for x in [1 .. n] do
+    for y in [1 .. n] do
+      Add(out[Color1Node(x)], Color2Node(x, y));
+      if x = y then
+        Add(out[Color2Node(x, y)], Widget(x));
+      fi;
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi]: Done

File: gap/attributes/isomorph.gi [hunk 15]
```diff
@@ -171,50 +190,150 @@ end);
...
+      Add(out[Color2Node(x, y)], Color3Node(x, y));
+      Add(out[Color3Node(x, y)], M[x][y]);
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/attributes/isomorph.gi]: Done

File: tst/standard/isomorph.tst (hunk 8 above)
Comment (james-d-mitchell) [reply tst/standard/isomorph.tst]: Done!

**Pull request merged**